[GAP-3] Mid-agreement payments (#24)
golemfactory/golem-architecture: gaps/gap-3_mid_agreement_payments/gap-3_mid_agreement_payments.md

## Motivation
Current payments scheme allows only to pay for tasks, after all computations are finished and Invoice is sent to Requestor.
In case of long-running tasks or services, Provider would expect, that he will be paid during
execution, because he doesn't want to wait for task finish.

## Specification

Setting `due_date` in Debit Note will cause yagna payment module to automatically pay for Debit Note,
after it was accepted. This specification describes, how Requestor and Provider should negotiate payments parameters.

### New properties

| Name | Type | Description |
| ---- | ---- | ----------- |
| `golem.com.scheme.payu.debit-note-interval-sec?` | Number (uint32) [Negotiable] | Minimum time (in seconds) between subsequent Debit Notes. The Provider must not issue the next Debit Note before the interval period. The Requestor ignores Debit Notes issued before interval elapses. |
| `golem.com.scheme.payu.payment-timeout-sec?` | Number (uint32) [Negotiable] | Maximum time to receive payment for any Debit Note. This time is counted from Debit Note’s issue date. |

- Requestor can ignore Debit Note, if it came before expected interval

>   - If Requestor is spammed with Debit Notes by Provider, he is allowed to break Agreement from Provider's fault

Reviewer: ```suggestion
  - If Requestor is spammed with Debit Notes by Provider, he is allowed to break Agreement stating (in TerminateAgreement) that the **reason** for termination is Provider's perceived misconduct. 
```

- In the mechanism suggested above a Debit Note shall have PaymentDueDate set, and the values must be set so that they correspond with the agreed `payment-interval-sec` value.
  - Intervals are relative to the agreement timestamp
- It is up to payment driver implementation to decide, when we can treat Debit Note as paid.
  - Requestor's payment driver must ensure, that transaction will be confirmed in given timestamp. 
  - Provider can check `DebitNoteSettledEvent` timestamp to know, if Debit Note was paid in time.

### Deprecation

`golem.com.scheme.payu.interval_sec` - Replaced by negotiable `golem.com.scheme.payu.debit-note.interval-sec?`. Provider
using mid-agreement payments specification can't set `interval_sec` property.

**Note:** We encourage using `golem.com.scheme.payu.debit-note.interval-sec?` property in new implementations,
even when parties don't need mid-agreement payments.

### Impact analysis

Following modifications need to be applied to components in Golem suite:

> - `yagna`

Reviewer 2: should we also include the modifications required for the "nagative paths" i.e. breaking the agreement due to failing to meet agreed times

Author: I added validation of time intervals on yapapi/yajsapi side.
I can't see any other cases that weren't covered

## Motivation
Current payments scheme allows only to pay for tasks, after all computations are finished and Invoice is sent to Requestor.
In case of long-running tasks or services, Provider would expect, that he will be paid during
execution, because he doesn't want to wait for task finish.

## Specification

Setting `due_date` in Debit Note will cause yagna payment module to automatically pay for Debit Note,
after it was accepted. This specification describes, how Requestor and Provider should negotiate payments parameters.

### New properties

| Name | Type | Description |
| ---- | ---- | ----------- |
| `golem.com.scheme.payu.debit-note-interval-sec?` | Number (uint32) [Negotiable] | Minimum time (in seconds) between subsequent Debit Notes. The Provider must not issue the next Debit Note before the interval period. The Requestor ignores Debit Notes issued before interval elapses. |
| `golem.com.scheme.payu.payment-timeout-sec?` | Number (uint32) [Negotiable] | Maximum time to receive payment for any Debit Note. This time is counted from Debit Note’s issue date. |

- Requestor can ignore Debit Note, if it came before expected interval

>   - If Requestor is spammed with Debit Notes by Provider, he is allowed to break Agreement stating (in TerminateAgreement) that the **Reason** for termination is Provider's perceived misconduct. He should set `golem.requestor.code` field in **Reason** structure to `ToManyDebitNotes`.

Reviewer: ```suggestion
  - If Requestor is spammed with Debit Notes by Provider, he is allowed to break Agreement stating (in TerminateAgreement) that the **Reason** for termination is Provider's perceived misconduct. He should set `golem.requestor.code` field in **Reason** structure to `TooManyDebitNotes`.
```

---
gap: 3
title: Mid Agreement payments
description: Specification of Agreement standard for negotiating and describing continues payments during task execution.  
author: nieznany.sprawiciel (@nieznanysprawiciel), stranger80 (@stranger80), shadeofblue (@shadeofblue)
status: Draft

> type: Feature

Author: @stranger80 Why it is feature, not standard?

Reviewer: To me it is a feature as it requires changes in components which are to implement this logic. A 'standard' amendment would only improve or publish standards but not provide 'features' in software.

| `golem.com.scheme.payu.debit-note-interval-sec?` | Number (uint32) [Negotiable] | Minimum time (in seconds) between subsequent Debit Notes. The Provider must not issue the next Debit Note before the interval period. The Requestor ignores Debit Notes issued before interval elapses. |
| `golem.com.scheme.payu.payment-timeout-sec?` | Number (uint32) [Negotiable] | Maximum time to receive payment for any Debit Note. This time is counted from Debit Note’s issue date. |

- Requestor can ignore Debit Note, if it came before expected interval
  - If Requestor is spammed with Debit Notes by Provider, he is allowed to break Agreement from Provider's fault
- In the mechanism suggested above a Debit Note shall have PaymentDueDate set, and the values must be set so that they correspond with the agreed `payment-interval-sec` value.
  - Intervals are relative to the agreement timestamp
- It is up to payment driver implementation to decide, when we can treat Debit Note as paid.
  - Requestor's payment driver must ensure, that transaction will be confirmed in given timestamp. 
  - Provider can check `DebitNoteSettledEvent` timestamp to know, if Debit Note was paid in time.

### Deprecation

> `golem.com.scheme.payu.interval_sec` - Replaced by negotiable `golem.com.scheme.payu.debit-note-interval-sec?`. Provider

Reviewer: ```suggestion
`golem.com.scheme.payu.interval_sec` - Replaced by negotiable `golem.com.scheme.payu.debit-note.interval-sec?`. Provider
```

using mid-agreement payments specification can't set `interval_sec` property.

> **Note:** We encourage using `golem.com.scheme.payu.debit-note-interval-sec?` property in new implementations,

Reviewer: ```suggestion
**Note:** We encourage using `golem.com.scheme.payu.debit-note.interval-sec?` property in new implementations,
```

even when parties don't need mid-agreement payments.

> 

Reviewer: ```suggestion
### Impact analysis

Following modifications need to be applied to components in Golem suite:

- `yagna`
  - payment module - confirm that DebitNote `due_date` logic is implemented to ensure accepted DebitNotes are paid in time
- `ya-provider` 
  - logic to support negotiation of `golem.com.scheme.payu.debit-note-interval-sec?` and `golem.com.scheme.payu.payment-timeout-sec?` properties
  - logic of `golem.com.scheme.payu.debit-note-interval-sec?` to ensure DebitNotes are generated based on negotiated time interval
  - logic of `golem.com.scheme.payu.payment-timeout-sec?` to validate payments are made according to negotiated terms
- `yapapi`/`yajsapi` 
  - logic to support negotiation of `golem.com.scheme.payu.debit-note-interval-sec?` and `golem.com.scheme.payu.payment-timeout-sec?` properties
  
  
```

## Rationale

### Debit Note interval and payment timeout relative to Agreement approve timestamp

Debit Note creation time can't serve as a base for payment timeout, because both Provider
and Requestor can lie about timestamps of this artifact. Moreover, Provider first issues Debit Note
and then sends it to Requestor.  

We considered malicious Provider, that tries to break Agreement, not from his fault, by waiting with
sending Debit Note, and stealing this way time required for making transaction on blockchain.

> The conclusion was, that we need absolut timestamp for computing payment timeout, which both parties

Reviewer: ```suggestion
The conclusion was, that we need absolute timestamp for computing payment timeout, which both parties
```

can agree on.

### Deprecation of `golem.com.scheme.payu.interval_sec`

> This property was used by Provider for declaring  interval between Debit Notes. Since more Debit Notes means

Reviewer: ```suggestion
This property was used by Provider for declaring the interval between Debit Notes. Since more Debit Notes means
```
